pkg/upgrade/control_plane.go

```diff
@@ -334,18 +287,31 @@ func (u *ControlPlaneUpgrader) updateMachines(machines *clusterapiv1alpha1.Machi
 			return err
 		}
 
-		if err := u.applyAnnotation(&machine); err != nil {
+		if err := u.applyAnnotation(newMachine); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
 func (u *ControlPlaneUpgrader) applyAnnotation(m *clusterapiv1alpha1.Machine) error {
+	original, err := json.Marshal(m)
+	if err != nil {
+		return errors.WithStack(err)
+	}
 	m.Annotations[UpgradeIDAnnotationKey] = u.upgradeID
```

Comment:
Do we need a nil check on the annotations map, with initialization if it's nil?

Reply:
no, ObjectMeta is not a pointer and neither is Annotations. But maps are kinda pointers by themselves so I ran this test: https://play.golang.org/p/wD2cRIV_vEY

We don't need one, but if you'd feel more comfortable with one, I can add it, easy enough.

Reply:
https://play.golang.org/p/dP6_hRTMbMY 😄

Reply:
hah, ok ok, i'll add a nil check